File: backend/src/main/java/com/codecool/tuttifrutti/service/UserService.java
package com.codecool.tuttifrutti.service;

import com.codecool.tuttifrutti.dto.RecipeDTO;
import com.codecool.tuttifrutti.dto.UserDTO;
import com.codecool.tuttifrutti.model.Ingredient;
import com.codecool.tuttifrutti.model.Recipe;
import com.codecool.tuttifrutti.model.User;
import com.codecool.tuttifrutti.repository.RecipeRepository;
import com.codecool.tuttifrutti.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class UserService {

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private RecipeRepository recipeRepository;

    @Autowired
    private BCryptPasswordEncoder bCryptPasswordEncoder;

    public User getUser(String username) {
        return this.userRepository.findByUsername(username);
    }

    public void likeRecipe(String username, long recipe_id) {
        User user = this.userRepository.findByUsername(username);
        if (user != null) {
            Optional<Recipe> recipe = this.recipeRepository.findById(recipe_id);
            if (recipe.isPresent() && user.getLikedRecipes().contains(recipe.get())) {
                user.removeLikedRecipe(recipe.get());
            } else {
                user.addLikedRecipe(recipe.get());
            }
            this.userRepository.save(user);
        }
    }

    public void addUsersRecipe(String username, RecipeDTO recipeDTO) {
        User user = this.userRepository.findByUsername(username);
        if (user!= null) {
            Recipe newRecipe = Recipe.builder()
                    .name(recipeDTO.getName())
                    .preparation(recipeDTO.getPreparation())
                    .imageUrl(recipeDTO.getImageUrl())
                    .difficulty(recipeDTO.getDifficulty())
                    .ingredients(recipeDTO.getIngredients()
                            .stream()
                            .map(ingredientDTO -> new Ingredient(ingredientDTO.getName(),
                                    ingredientDTO.getAmount(),
                                    ingredientDTO.getUnit()))
                            .collect(Collectors.toList()))
                    .user(user)
                    .build();

            List<Ingredient> newIngredients = new ArrayList<>();
            recipeDTO.getIngredients()
                    .forEach(ingredientDTO -> newIngredients.add(
                            new Ingredient(ingredientDTO.getName(),
                                    ingredientDTO.getAmount(),
                                    ingredientDTO.getUnit(),
                                    newRecipe)));
            newRecipe.setIngredients(newIngredients);

            user.getRecipes().add(newRecipe);
            newRecipe.setUser(user);
//            userObject.addNewRecipe(recipeDTO);
//            this.recipeRepository.save(newRecipe);
            this.userRepository.save(user);
        }
    }

    public void editUsersRecipe(String username, long recipe_id, Recipe recipe) {
        User user = this.userRepository.findByUsername(username);
        if (user != null) {
            Recipe oldRecipe = user.getRecipes().stream()
                    .filter(currentRecipe -> currentRecipe.getId() == recipe_id)
                    .findFirst().get();
            oldRecipe = recipe;
            oldRecipe.setUser(user);
            this.recipeRepository.save(recipe);
//            this.userRepository.save(userObject);
        }
    }

    public void deleteUsersRecipe(String username, long recipe_id) {
        User user = this.userRepository.findByUsername(username);
        ArrayList<Recipe> recipes = new ArrayList<Recipe>();
        if (user != null) {
            for (Recipe recipe : user.getRecipes()) {
                if (recipe.getId() == recipe_id) {
                    recipes.add(recipe);
                }
            }
            user.getRecipes().remove(recipes.get(0));
            this.userRepository.save(user);

            Optional<Recipe> recipe = this.recipeRepository.findById(recipe_id);
            recipe.ifPresent(recipe1 -> this.recipeRepository.delete(recipe1));
        }
    }

    public void registration(UserDTO userDTO) {
        User newUser = User.builder()
                .username(userDTO.getUsername())
                .password(this.bCryptPasswordEncoder.encode(userDTO.getPassword()))
//                .registrationDate(userDTO.getRegistrationDate())
                .registrationDate(LocalDate.now())
                .build();
        this.userRepository.save(newUser);
    }

    public User login(String username, String password) {
        User user = this.userRepository.findByUsername(username);
        System.out.println(user != null && user.getPassword().equals(password));
        if (user != null && user.getPassword().equals(password)) return user;

        return null;
    }
}

File: frontend/src/app/services/user.service.ts
import { Injectable } from '@angular/core';
import { Ingredient } from '../model/ingredient.model';
import { HttpClient, HttpResponse } from '@angular/common/http';
import { Recipe } from '../model/recipe.model';
import { User } from '../model/user.model';
import { AuthService } from './auth.service';
import { Router } from '@angular/router';

@Injectable()
export class UserService {
  baseUsersUrl: string = "http://localhost:8080/users";
  user: User;

  constructor(private http: HttpClient,
              private authService:AuthService,
              private router: Router){}

  addRecipe(recipeDTO: { name: string;
    preparation: string;
    imageUrl: string;
    difficulty: string;
    ingredients: Ingredient[] }) {
    this.http.post<any>(`${this.baseUsersUrl}/${this.user.id}/add-recipe`, recipeDTO).subscribe();
  }

  login(username: string, password: string) {
    return this.http.post<User>(`${this.baseUsersUrl}/login`,
      {username: username, password: password},
      {observe: 'response'}).subscribe(
      (response: HttpResponse<any>) => {
        if (response.headers.get('Authorization')) {
          this.authService.saveToken(response.headers.get('Authorization'));
          // this.checkIsUserValid();
        }
      }
    );
    }

  addLike(recipeId: number) {
    this.http.put<any>(`${this.baseUsersUrl}/${this.user.id}/liked-recipes/recipes/${recipeId}`, null)
      .subscribe();
  }

  register(userDTO: { username: string,
    password: string,
    registrationDate: string,
    recipes: Recipe[],
    likedRecipes: Recipe[]  }) {
    this.http.post<any>(`${this.baseUsersUrl}/registration`, userDTO).subscribe();
    this.router.navigate(['login']);

  }

  checkIsUserValid(){
    if (this.authService.loggedIn()) {
      this.authService.isLoggedIn.next(true);
      this.router.navigate(['/']);
    } else  {
      alert("invalid username or password");
    }
  }

  editUsersRecipe(username: string, recipeId: number, recipe: Recipe) {
    this.http.put(`${this.baseUsersUrl}/${username}/recipes/${recipeId}`, recipe);
  }

  deleteUsersRecipe(username: string, recipeId: number) {
    return this.http.delete<Recipe>(`${this.baseUsersUrl}/${username}/recipes/${recipeId}`).subscribe();
  }

  getUser(username: string) {
    return this.http.get<User>(`${this.baseUsersUrl}/${username}`);
  }

}
